From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mia <vottus@vott.us>
Date: Wed, 12 May 2021 06:34:05 +0200
Subject: [PATCH] fix: completly implement ayaka and fix bugs


diff --git a/src/main/java/eu/matherion/api/servers/ServerStateUpdate.java b/src/main/java/eu/matherion/api/servers/ServerStateUpdate.java
index d165ec4fd35b7cd2e278fdaa707eea61305d7fa1..11aabf45190d23ed09123542a93073da3bf658c5 100644
--- a/src/main/java/eu/matherion/api/servers/ServerStateUpdate.java
+++ b/src/main/java/eu/matherion/api/servers/ServerStateUpdate.java
@@ -1,18 +1,24 @@
 package eu.matherion.api.servers;
 
+import com.google.gson.annotations.Expose;
 import eu.matherion.api.servers.ServerState;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
 public class ServerStateUpdate {
 
-    private final String serverName;
+    @Expose
+    public final String serverName;
 
     @Nullable
-    private final ServerState serverState;
+    @Expose
+    public final ServerState serverState;
 
     public ServerStateUpdate(String serverName, @Nullable ServerState serverState) {
         this.serverName = serverName;
         this.serverState = serverState;
+        System.out.println("New ServerUpdate");
+        System.out.println(serverName);
+        System.out.println(serverState);
     }
 
     public String getServerName() {
