From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Adam Volkman <itismaku@gmail.com>
Date: Fri, 14 May 2021 21:51:45 +0200
Subject: [PATCH] fix: fixed FetchAllStatesResponse (fixed by ZorTik :-))


diff --git a/src/main/java/eu/matherion/server/ayaka/Ayaka.java b/src/main/java/eu/matherion/server/ayaka/Ayaka.java
index a26b497cf15eeb8e90d8f996951158f0e5d0c686..b3f18bbed94e2ddfc800e4647599cef3e97825fa 100644
--- a/src/main/java/eu/matherion/server/ayaka/Ayaka.java
+++ b/src/main/java/eu/matherion/server/ayaka/Ayaka.java
@@ -168,13 +168,12 @@ public class Ayaka implements ServerPubSub {
   public Map<String, ServerState> getAllStates(FetchAllStatesData data) {
     GetRequest req = restClient.get("/state/fetchAll");
 
-    if (data.name != null)
+    if (data.name != null) {
       req.queryString(Map.of("name", data.name));
-
+    }
     try {
       RestResponse<FetchAllStatesResponse> response = new RestResponse<>(req.asJson().getBody().getObject(), FetchAllStatesResponse.class);
       if (!response.success) return null;
-
       return response.data.states;
     } catch (Exception e) {
       e.printStackTrace();
